lib/active_model/serializer/association.rb

```diff
@@ -10,24 +10,111 @@ class Serializer
+      # build serializer's association by its reflection.
+      #
+      # @param [Serializer] parent_serializer for given association
+      # @param [Hash{Symbol => Object}] parent_serializer_options
+      #
+      # @example
+      #    # Given the following serializer defined:
+      #    class PostSerializer < ActiveModel::Serializer
+      #      has_many :comments, serializer: CommentSummarySerializer
+      #    end
+      #
+      #    # Then you instantiate your serializer
+      #    post_serializer = PostSerializer.new(post, foo: 'bar') #
+      #    # to build association for comments you need to get reflection
+      #    comments_reflection = PostSerializer._reflections.detect { |r| r.name == :comments }
+      #    # and #build_association
+      #    comments_reflection.build_association(post_serializer, foo: 'bar')
+      #
+      # @api private
+      def self.from_reflection(reflection, association_options)
+        block = nil
+        Association.new(reflection.name, association_options.merge(reflection: reflection), block)
+      end
+
+      attr_reader :parent_serializer, :parent_serializer_options, :include_slice, :reflection, :association_options
+      def initialize(*)
+        super
+        @association_options = options
+        @parent_serializer = association_options.fetch(:parent_serializer)
+        @parent_serializer_options = association_options.fetch(:parent_serializer_options)
+        @include_slice = association_options.fetch(:include_slice)
+        @reflection = association_options.fetch(:reflection)
+
+        self.options = reflection.options.dup
+        # Pass the parent's namespace onto the child serializer
+        options[:namespace] ||= parent_serializer_options[:namespace]
+        serializer
+      end
+
       # @return [Symbol]
       def key
         options.fetch(:key, name)
       end
 
-      # @return [ActiveModel::Serializer, nil]
-      def serializer
-        options[:serializer]
-      end
-
       # @return [Hash]
       def links
         options.fetch(:links) || {}
       end
 
       # @return [Hash, nil]
       def meta
-        options[:meta]
+        reflection.options[:association_meta]
+      end
+
+      def object
+        @object ||= reflection.value(parent_serializer, include_slice)
+      end
+
+      def include_data?
+        reflection.include_data?(include_slice)
+      end
+
+      # @return [ActiveModel::Serializer, nil]
+      def serializer
+        return @serializer if defined?(@serializer)
+        if serializer_class
+          serialize_object!(object)
+        elsif !object.nil? && !object.instance_of?(Object)
+          options[:virtual_value] = object
+        end
+        @serializer = options[:serializer]
+      end
+
+      def serialize_object!(object)
+        serializer = instantiate_serializer(object)
+        if serializer.nil?
+          options[:virtual_value] = object.try(:as_json) || object
```

Comment:
BUG originates here per https://github.com/rails-api/active_model_serializers/issues/2027#issuecomment-272570471 that an association without a serializer should NOT be serialized like this for JSON API, since that has a required format.